// @medusajs/modules-sdk/dist/remote-query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteQuery = void 0;
const medusa_module_1 = require("./medusa-module");
const orchestration_1 = require("@medusajs/orchestration");
const utils_1 = require("@medusajs/utils");
class RemoteQuery {
    constructor(modulesLoaded, remoteFetchData) {
        this.modulesMap = new Map();
        if (!modulesLoaded?.length) {
            modulesLoaded = medusa_module_1.MedusaModule.getLoadedModules().map((mod) => Object.values(mod)[0]);
        }
        const servicesConfig = [];
        for (const mod of modulesLoaded) {
            if (!mod.__definition.isQueryable) {
                continue;
            }
            if (this.modulesMap.has(mod.__definition.key)) {
                throw new Error(`Duplicated instance of module ${mod.__definition.key} is not allowed.`);
            }
            this.modulesMap.set(mod.__definition.key, mod);
            servicesConfig.push(mod.__joinerConfig);
        }
        this.remoteJoiner = new orchestration_1.RemoteJoiner(servicesConfig, remoteFetchData ?? this.remoteFetchData.bind(this));
    }
    setFetchDataCallback(remoteFetchData) {
        this.remoteJoiner.setFetchDataCallback(remoteFetchData);
    }
    static getAllFieldsAndRelations(data, prefix = "") {
        let fields = new Set();
        let relations = [];
        data.fields?.forEach((field) => {
            fields.add(prefix ? `${prefix}.${field}` : field);
        });
        if (data.expands) {
            for (const property in data.expands) {
                const newPrefix = prefix ? `${prefix}.${property}` : property;
                relations.push(newPrefix);
                fields.delete(newPrefix);
                const result = RemoteQuery.getAllFieldsAndRelations(data.expands[property], newPrefix);
                result.select.forEach(fields.add, fields);
                relations = relations.concat(result.relations);
            }
        }
        return { select: [...fields], relations };
    }
    hasPagination(options) {
        if (!options) {
            return false;
        }
        const attrs = ["skip", "cursor"];
        return Object.keys(options).some((key) => attrs.includes(key));
    }
    buildPagination(options, count) {
        return {
            skip: options.skip,
            take: options.take,
            cursor: options.cursor,
            // TODO: next cursor
            count,
        };
    }
    async remoteFetchData(expand, keyField, ids, relationship) {
        const serviceConfig = expand.serviceConfig;
        const service = this.modulesMap.get(serviceConfig.serviceName);
        let filters = {};
        const options = {
            ...RemoteQuery.getAllFieldsAndRelations(expand),
        };
        const availableOptions = [
            "skip",
            "take",
            "limit",
            "offset",
            "cursor",
            "sort",
        ];
        const availableOptionsAlias = new Map([
            ["limit", "take"],
            ["offset", "skip"],
        ]);
        for (const arg of expand.args || []) {
            if (arg.name === "filters" && arg.value) {
                filters = { ...arg.value };
            }
            else if (availableOptions.includes(arg.name)) {
                const argName = availableOptionsAlias.has(arg.name)
                    ? availableOptionsAlias.get(arg.name)
                    : arg.name;
                options[argName] = arg.value;
            }
        }
        if (ids) {
            filters[keyField] = ids;
        }
        const hasPagination = this.hasPagination(options);
        let methodName = hasPagination ? "listAndCount" : "list";
        if (relationship?.args?.methodSuffix) {
            methodName += (0, utils_1.toPascalCase)(relationship.args.methodSuffix);
        }
        else if (serviceConfig?.args?.methodSuffix) {
            methodName += (0, utils_1.toPascalCase)(serviceConfig.args.methodSuffix);
        }
        if (typeof service[methodName] !== "function") {
            throw new Error(`Method "${methodName}" does not exist on "${serviceConfig.serviceName}"`);
        }
        const result = await service[methodName](filters, options);
        if (hasPagination) {
            const [data, count] = result;
            return {
                data: {
                    rows: data,
                    metadata: this.buildPagination(options, count),
                },
                path: "rows",
            };
        }
        return {
            data: result,
        };
    }
    async query(query, variables = {}) {
        return await this.remoteJoiner.query(orchestration_1.RemoteJoiner.parseQuery(query, variables));
    }
}
exports.RemoteQuery = RemoteQuery;
