// @medusajs/utils/dist/modules-sdk/load-module-database-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadDatabaseConfig = void 0;
var common_1 = require("../common");
function getEnv(key, moduleName) {
    var _a;
    var value = (_a = process.env["".concat(moduleName.toUpperCase(), "_").concat(key)]) !== null && _a !== void 0 ? _a : process.env["".concat(key)];
    return value !== null && value !== void 0 ? value : "";
}
function isModuleServiceInitializeOptions(obj) {
    return !!(obj === null || obj === void 0 ? void 0 : obj.database);
}
function getDefaultDriverOptions(clientUrl) {
    var _a, _b;
    var localOptions = {
        connection: {
            ssl: false,
        },
    };
    var remoteOptions = {
        connection: {
            ssl: {
                rejectUnauthorized: false,
            },
        },
    };
    if (clientUrl) {
        return clientUrl.match(/localhost/i) ? localOptions : remoteOptions;
    }
    return ((_a = process.env.NODE_ENV) === null || _a === void 0 ? void 0 : _a.match(/prod/i))
        ? remoteOptions
        : ((_b = process.env.NODE_ENV) === null || _b === void 0 ? void 0 : _b.match(/dev/i))
            ? localOptions
            : {};
}
/**
 * Load the config for the database connection. The options can be retrieved
 * e.g through PRODUCT_* (e.g PRODUCT_POSTGRES_URL) or * (e.g POSTGRES_URL) environment variables or the options object.
 * @param options
 * @param moduleName
 */
function loadDatabaseConfig(moduleName, options) {
    var _a, _b, _c, _d, _e, _f, _g;
    var clientUrl = getEnv("POSTGRES_URL", moduleName);
    var database = {
        clientUrl: getEnv("POSTGRES_URL", moduleName),
        schema: (_a = getEnv("POSTGRES_SCHEMA", moduleName)) !== null && _a !== void 0 ? _a : "public",
        driverOptions: JSON.parse(getEnv("POSTGRES_DRIVER_OPTIONS", moduleName) ||
            JSON.stringify(getDefaultDriverOptions(clientUrl))),
        debug: (_c = (_b = process.env.NODE_ENV) === null || _b === void 0 ? void 0 : _b.startsWith("dev")) !== null && _c !== void 0 ? _c : false,
    };
    if (isModuleServiceInitializeOptions(options)) {
        database.clientUrl = (_d = options.database.clientUrl) !== null && _d !== void 0 ? _d : database.clientUrl;
        database.schema = (_e = options.database.schema) !== null && _e !== void 0 ? _e : database.schema;
        database.driverOptions =
            (_f = options.database.driverOptions) !== null && _f !== void 0 ? _f : getDefaultDriverOptions(database.clientUrl);
        database.debug = (_g = options.database.debug) !== null && _g !== void 0 ? _g : database.debug;
    }
    if (!database.clientUrl) {
        throw new common_1.MedusaError(common_1.MedusaError.Types.INVALID_ARGUMENT, "No database clientUrl provided. Please provide the clientUrl through the PRODUCT_POSTGRES_URL or POSTGRES_URL environment variable or the options object in the initialize function.");
    }
    return database;
}
exports.loadDatabaseConfig = loadDatabaseConfig;
